cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -0,0 +1,492 @@
+          } else if (decimal_precision <= MAX_DECIMAL64_PRECISION) {
+            gpuOutputByteArrayAsInt(ptr, len, static_cast<int64_t*>(dst));
+          } else {
+            gpuOutputByteArrayAsInt(ptr, len, static_cast<__int128_t*>(dst));
+          }
+        } else {
+          gpuOutputString(s, sb, src_pos, dst);
+        }
+      } else if (dtype == BOOLEAN) {
+        gpuOutputBoolean(sb, src_pos, static_cast<uint8_t*>(dst));
+      } else if (s->col.converted_type == DECIMAL) {
+        switch (dtype) {
+          case INT32: gpuOutputFast(s, sb, src_pos, static_cast<uint32_t*>(dst)); break;
+          case INT64: gpuOutputFast(s, sb, src_pos, static_cast<uint2*>(dst)); break;
+          default:
+            if (s->dtype_len_in <= sizeof(int32_t)) {
+              gpuOutputFixedLenByteArrayAsInt(s, sb, src_pos, static_cast<int32_t*>(dst));
+            } else if (s->dtype_len_in <= sizeof(int64_t)) {
+              gpuOutputFixedLenByteArrayAsInt(s, sb, src_pos, static_cast<int64_t*>(dst));
+            } else {
+              gpuOutputFixedLenByteArrayAsInt(s, sb, src_pos, static_cast<__int128_t*>(dst));
+            }
+            break;
+        }
+      } else if (dtype == INT96) {
+        gpuOutputInt96Timestamp(s, sb, src_pos, static_cast<int64_t*>(dst));
+      } else if (dtype_len == 8) {
+        if (s->dtype_len_in == 4) {
+          // Reading INT32 TIME_MILLIS into 64-bit DURATION_MILLISECONDS
+          // TIME_MILLIS is the only duration type stored as int32:
+          // https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#deprecated-time-convertedtype
+          gpuOutputFast(s, sb, src_pos, static_cast<uint32_t*>(dst));
+        } else if (s->ts_scale) {
+          gpuOutputInt64Timestamp(s, sb, src_pos, static_cast<int64_t*>(dst));
+        } else {
+          gpuOutputFast(s, sb, src_pos, static_cast<uint2*>(dst));
+        }
+      } else if (dtype_len == 4) {
+        gpuOutputFast(s, sb, src_pos, static_cast<uint32_t*>(dst));
+      } else {
+        gpuOutputGeneric(s, sb, src_pos, static_cast<uint8_t*>(dst), dtype_len);
+      }
+    }
+
+    pos += batch_size;
+  }
+}
+
+/**
+ * @brief Kernel for co the column data stored in the pages
+ *
+ * This function will write the page data and the page data's validity to the
+ * output specified in the page's column chunk. If necessary, additional
+ * conversion will be performed to translate from the Parquet datatype to
+ * desired output datatype (ex. 32-bit to 16-bit, string to hash).
+ *
+ * @param pages List of pages
+ * @param chunks List of column chunks
+ * @param min_row Row index to start reading at
+ * @param num_rows Maximum number of rows to read
+ */
+template <typename level_t>
+__global__ void __launch_bounds__(decode_block_size) gpuDecodePageDataFixed(
+  PageInfo* pages, device_span<ColumnChunkDesc const> chunks, size_t min_row, size_t num_rows)
+{
+  __shared__ __align__(16) page_state_s state_g;
```

Comment:
Could we gain anything by splitting `page_state_s` so that only the relevant members are included in the specialized kernels?
I assume its size is small compared to `rle_run` objects so it might not move the needle WRT shared memory use, but asking just in case.

Reply:
Yes, I think so. We should add this to another follow on for us to tackle in the future. Will file. @nvdbaranec fyi.

Reply:
Filed this: https://github.com/rapidsai/cudf/issues/15267